### DISCUSSION TITLE
Ubuntu (v22.04.2 LTS) nginx (v1.18.0) 404 error when trying to get JavaScript file for Plausible

### DISCUSSION THREAD

**Discussion by TomTheTech01** [plausible/analytics#3213] [Self-Hosted Support]
Hello,

I recently launched a website on a VPS running Ubuntu (v22.04.2 LTS) using nginx (v1.18.0) as a web server. I used certbot to get an SSL cert.

I'm trying to self host Plausible. I'm able to access the Plausible dashboard by going to http://tomkdance.com:8000/.

When I get to the dashboard, I see a "Waiting for first pageview for tomkdance.com" message and a small green circle (see attachment).

![01 waiting for first pageview](https://github.com/plausible/analytics/assets/141073821/05823e88-c818-4fe2-9f98-7a1f40ba14b0)

I'm using the following JavaScript snippet (see attachments).
<script defer data-domain="tomkdance.com" src="https://tomkdance.com/js/script.js"></script>

![02 add javascript snippet](https://github.com/plausible/analytics/assets/141073821/eb2f4312-2871-42db-9f32-9cdf1dda5bf4)
![03 index html](https://github.com/plausible/analytics/assets/141073821/f8d47e2d-64a0-4bc7-8fc9-4bb5a42ee4fa)

I disabled my popup blocker for testing. I opened up DevTools in Chrome and noticed that there's a 404 error when trying to get the JavaScript file.

I also noticed that the remote address ends in :443.

![04 404 error detail](https://github.com/plausible/analytics/assets/141073821/f069b7bc-b250-420f-b47d-bb739fc6a01a)

I have not tried proxying plausible through nginx yet since I was under the impression I didn't need to do that (but I'm open to it if that's the easiest way forward).

If anyone is willing to help I'm happy to provide whatever information is necessary. Thank you!

**Comment by ruslandoga** [comment]
👋 @TomTheTech01 

Can you please try using `http://tomkdance.com:8000/js/script.js` for `src` in script?

**Comment by TomTheTech01** [comment]
@ruslandoga Thank you for the quick response!

Here's what happened after I made the change you suggested.

![script change](https://github.com/plausible/analytics/assets/141073821/1ddfad97-4440-4a6f-a9b3-9b90385cbe54)

**Reply by ruslandoga** [reply]
Ah, right. Chrome forces to use https. The easiest way to use https is probably Cloudflare or Caddy.

**Reply by ruslandoga** [reply]
[Here's](https://github.com/plausible/hosting/tree/master/reverse-proxy#no-existing-reverse-proxy) how to add Caddy to the default setup.

**Comment by TomTheTech01** [comment]
When I entered the command shown on that page for caddy, I got the following output. It wasn't clear to me how to continue. It appears that caddy is a web server? I'm unfamiliar with it and came across quite a few articles for setting up a reverse proxy on nginx during troubleshooting, so I'll try that next since I'm more familiar with nginx.

```
$ docker-compose -f docker-compose.yml -f reverse-proxy/docker-compose.caddy-gen.yml up
$: command not found
```

**Reply by ruslandoga** [reply]
`$` indicates the command line prompt. It should not be inputted. If `docker-compose` is not found either, try using `docker compose`. Delegating TLS cert generation and termination to Caddy is quite a bit easier than doing it manually with nginx. Yes, both Caddy and nginx are webservers.

**Comment by TomTheTech01** [comment]
Update - I took ruslandoga's advice to use caddy and ended up rebuilding my Digital Ocean droplet to use caddy instead of nginx. SSL is working with the default caddy HTML page, now I just have to upload my files and go through the plausible tutorial again.

**Comment by TomTheTech01** [comment]
@ruslandoga 
My site is up and SSL secured via caddy.
![site secured](https://github.com/plausible/analytics/assets/141073821/c71cd205-5d12-4473-a72e-a93cf536e65d)

Here's my HTML header.
![HTML header](https://github.com/plausible/analytics/assets/141073821/86fb167a-11b9-4a78-95a4-c0e159edd6b3)

When I go to tomkdance.com:8000 in Chrome and click on my site, I'm back to "Waiting for first pageview".

I'll retrace my steps tomorrow morning, but do you have any advice please?

**Comment by dimejo** [comment]
I think the link in your site needs to point to `tomkdance.com:8000` otherwise it will try to load the script from your website, not from Plausible.
`<script defer data-domain="tomkdance.com" src="https://tomkdance.com:8000/js/script.js"></script>`

Is there any reason why you aren't using a subdomain for Plausible? This would make things a lot easier and all it takes is a `CNAME` entry in your DNS.

**Reply by TomTheTech01** [reply]
@dimejo Thanks for your response. I tried your suggestion and got this result.
![dimejo](https://github.com/plausible/analytics/assets/141073821/168ec932-e8d4-4d59-9336-7a2766ba8cd4)

The reason I'm not using a subdomain for Plausible is because it wasn't in this documentation that I followed.
https://plausible.io/docs/self-hosting

Is there a better source?

**Reply by TomTheTech01** [reply]
Here's the recommended snippet from the plausible dashboard.
![snippet](https://github.com/plausible/analytics/assets/141073821/8ee012b8-48cb-4682-8fd5-2e50b9990d13)

**Reply by TomTheTech01** [reply]
I switched back to the recommended snippet and got the following console error when loading my site.
![error](https://github.com/plausible/analytics/assets/141073821/ef6d831e-47bd-48e0-94f8-69a170b91a77)

**Reply by ruslandoga** [reply]
👋 @TomTheTech01 

The easiest way is to use a separate (sub)domain name for Plausible, just as @dimejo recommended. Otherwise you'd need to setup Caddy to be able to distinguish between "Plausible" paths and your website's paths on the same domain which is very error prone.

**Reply by TomTheTech01** [reply]
@ruslandoga Thank you. Can you point me towards any documentation to do that?

**Reply by ruslandoga** [reply]
- You'd need to setup the subdomain on your DNS provider, like Cloudflare or AWS Route53 with either an A or CNAME records.
- Then you'd add that subdomain [here]( https://github.com/plausible/hosting/blob/44925a9937cd9a226417b61acff945549715dbd5/reverse-proxy/docker-compose.caddy-gen.yml#L18)
- Then you'd need to remember to include that config when you start your containers: 

```console
$ docker compose -f docker-compose.yml -f reverse-proxy/docker-compose.caddy-gen.yml up
```

**Reply by TomTheTech01** [reply]
I tried setting up the subdomain like this.
![create new record](https://github.com/plausible/analytics/assets/141073821/b9888159-691a-48e9-a0b6-ae6ac7506cb5)

Where can I find docker-compose.caddy-gen.yml again please? 

There's a reverse proxy mentioned in your link, I haven't set that up if it's a requirement.

**Reply by ruslandoga** [reply]
Here's some info on using Caddy as a reverse proxy in an automated fashion:

- https://github.com/plausible/hosting/blob/master/reverse-proxy/docker-compose.caddy-gen.yml
- https://plausible.io/docs/self-hosting#2-reverse-proxy

**Reply by ruslandoga** [reply]
If you are using Caddy directly on the same host that your website's static files and Plausible are, you can check their documentation on how to tie it all up.

It'd probably be a combination of these three articles:

- https://caddyserver.com/docs/quick-starts/static-files - to serve your website (as I understand it's currently static)
- https://caddyserver.com/docs/quick-starts/reverse-proxy - to redirect Plausible requests to a running Plausible container
- https://caddyserver.com/docs/caddyfile-tutorial - on how to express it all in a Caddyfile, it'd probably look like this:

```caddyfile
tomkdance.com {
    root * /path/to/your/tomkdance.com/files
    file_server
}

plausible.tomkdance.com {
    reverse_proxy localhost:8000
}
```

**Reply by TomTheTech01** [reply]
I'm having trouble locating docker-compose.caddy-gen.yml even after numerous google searches.
![docker-compose caddy-gen yml](https://github.com/plausible/analytics/assets/141073821/69903f87-08b7-4751-9759-12ea3c75a9c8)

I'm running Ubuntu 22.04 and it's not in /etc/docker.

**Reply by ruslandoga** [reply]
It should be cloned together with all the other configuration files in this step: https://plausible.io/docs/self-hosting#1-clone-the-hosting-repo

**Reply by TomTheTech01** [reply]
@ruslandoga Thank you for your patience.

1. setup subdomain
![dns records](https://github.com/plausible/analytics/assets/141073821/e0ad9764-7f2c-44c8-8f9c-fcf5ff4b2fad)

2. added subdomain to caddy-gen.yml
![caddy-gen-yml](https://github.com/plausible/analytics/assets/141073821/fdadb868-d7ef-4bf2-abab-f26eb649fc9f)

3. ran the following command
$ docker compose -f docker-compose.yml -f reverse-proxy/docker-compose.caddy-gen.yml up
![docker compose](https://github.com/plausible/analytics/assets/141073821/2d2a094a-e07d-42f1-98d5-c6f9ff4c64f9)

4. stopped and started caddy and docker services

I restored my droplet and reinstalled plausible, so I haven't tried anything after the reply that I'm quoting.

**Reply by TomTheTech01** [reply]
Woops, this is what I was trying to reply to.

> * You'd need to setup the subdomain on your DNS provider, like Cloudflare or AWS Route53 with either an A or CNAME records.
> * Then you'd add that subdomain [here](https://github.com/plausible/hosting/blob/44925a9937cd9a226417b61acff945549715dbd5/reverse-proxy/docker-compose.caddy-gen.yml#L18)
> * Then you'd need to remember to include that config when you start your containers:
> 
> ```
> $ docker compose -f docker-compose.yml -f reverse-proxy/docker-compose.caddy-gen.yml up
> ```

**Comment by dimejo** [comment]
Unfortunately I have no idea why your setup doesn't work and I don't have any experience running Caddy in a Docker container. For me the most simple setup (used it numerous time) is installing Caddy via package manager and providing relevant configuration in Caddyfile.

[Installing Caddy on Debian, Ubuntu or Raspbian:](https://caddyserver.com/docs/install#debian-ubuntu-raspbian)
```
sudo apt install -y debian-keyring debian-archive-keyring apt-transport-https
curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/gpg.key' | sudo gpg --dearmor -o /usr/share/keyrings/caddy-stable-archive-keyring.gpg
curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/debian.deb.txt' | sudo tee /etc/apt/sources.list.d/caddy-stable.list
sudo apt update
sudo apt install caddy
```
`/etc/caddy/Caddyfile`:
```
{
    email me@example.net
}

tomkdance.com {
    root * /path/to/your/files
    file_server
}

plausible.tomkdance.com {
    # Reverse proxy all requests to plausible container:
    reverse_proxy localhost:8000
}
```

Further steps:
* Adjust the `BASE_URL` in `docker-compose.yml` or `plausible-conf.env` to `plausible.tomkdance.com`
* Make sure that no other service is using ports 80 and 443
* Check that ports 80 and 443 in firewall are open
* Start Plausible container
* Start Caddy e.g. by running `systemctl enable --now caddy.service`
* Visit plausible.tomkdance.com to see if Plausible is working
* Adjust the script on tomkdance.com to point to `https://plausible.tomkdance.com/js/script.js`

Please adjust where necessary because most of it is from memory.

PS: You should lower the TTL on your DNS record to something more reasonable in case you need to change the record and don't want to wait 12 hours for it to take effect.

**Reply by TomTheTech01** [reply]
Thank you for the response. I rolled my droplet back to a Digital Ocean snapshot where SSL was working but before I installed docker and plausible.

My site loaded, and then I added the plausible.tomkdance.com block that you suggested. Here is what the caddy file looks like now. I'm going to reinstall plausible and see how it goes, but if anyone sees a problem with my caddy file please let me know.

![caddy file](https://github.com/plausible/analytics/assets/141073821/26478f6a-eccf-4603-b228-574705dd40fe)
